test/first_party/src/common/nanoprintf.h

```diff
@@ -0,0 +1,1333 @@
+#define NANOPRINTF_USE_FLOAT_FORMAT_SPECIFIERS 1
+#define NANOPRINTF_USE_LARGE_FORMAT_SPECIFIERS 0
+#define NANOPRINTF_USE_BINARY_FORMAT_SPECIFIERS 0
+#define NANOPRINTF_USE_WRITEBACK_FORMAT_SPECIFIERS 0
+#endif
+
+// If anything's been configured, everything must be configured.
+#ifndef NANOPRINTF_USE_FIELD_WIDTH_FORMAT_SPECIFIERS
+#error NANOPRINTF_USE_FIELD_WIDTH_FORMAT_SPECIFIERS must be #defined to 0 or 1
+#endif
+#ifndef NANOPRINTF_USE_PRECISION_FORMAT_SPECIFIERS
+#error NANOPRINTF_USE_PRECISION_FORMAT_SPECIFIERS must be #defined to 0 or 1
+#endif
+#ifndef NANOPRINTF_USE_FLOAT_FORMAT_SPECIFIERS
+#error NANOPRINTF_USE_FLOAT_FORMAT_SPECIFIERS must be #defined to 0 or 1
+#endif
+#ifndef NANOPRINTF_USE_LARGE_FORMAT_SPECIFIERS
+#error NANOPRINTF_USE_LARGE_FORMAT_SPECIFIERS must be #defined to 0 or 1
+#endif
+#ifndef NANOPRINTF_USE_BINARY_FORMAT_SPECIFIERS
+#error NANOPRINTF_USE_BINARY_FORMAT_SPECIFIERS must be #defined to 0 or 1
+#endif
+#ifndef NANOPRINTF_USE_WRITEBACK_FORMAT_SPECIFIERS
+#error NANOPRINTF_USE_WRITEBACK_FORMAT_SPECIFIERS must be #defined to 0 or 1
+#endif
+
+// Ensure flags are compatible.
+#if (NANOPRINTF_USE_FLOAT_FORMAT_SPECIFIERS == 1)                              \
+    && (NANOPRINTF_USE_PRECISION_FORMAT_SPECIFIERS == 0)
+#error Precision format specifiers must be enabled if float support is enabled.
+#endif
+
+// intmax_t / uintmax_t require stdint from c99 / c++11
+#if NANOPRINTF_USE_LARGE_FORMAT_SPECIFIERS == 1
+#ifndef _MSC_VER
+#ifdef __cplusplus
+#if __cplusplus < 201103L
+#error large format specifier support requires C++11 or later.
+#endif
+#else
+#if __STDC_VERSION__ < 199409L
+#error nanoprintf requires C99 or later.
+#endif
+#endif
+#endif
+#endif
+
+// Figure out if we can disable warnings with pragmas.
+#ifdef __clang__
+#define NANOPRINTF_CLANG 1
+#define NANOPRINTF_GCC_PAST_4_6 0
+#else
+#define NANOPRINTF_CLANG 0
+#if defined(__GNUC__)                                                          \
+    && ((__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ > 6)))
+#define NANOPRINTF_GCC_PAST_4_6 1
+#else
+#define NANOPRINTF_GCC_PAST_4_6 0
+#endif
+#endif
+
+#if NANOPRINTF_CLANG || NANOPRINTF_GCC_PAST_4_6
+#define NANOPRINTF_HAVE_GCC_WARNING_PRAGMAS 1
+#else
+#define NANOPRINTF_HAVE_GCC_WARNING_PRAGMAS 0
+#endif
+
+#if NANOPRINTF_HAVE_GCC_WARNING_PRAGMAS
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+#pragma GCC diagnostic ignored "-Wimplicit-fallthrough"
+#ifdef __cplusplus
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+#endif
+#pragma GCC diagnostic ignored "-Wpadded"
+#pragma GCC diagnostic ignored "-Wfloat-equal"
+#if NANOPRINTF_CLANG
+#pragma GCC diagnostic ignored "-Wc++98-compat-pedantic"
+#pragma GCC diagnostic ignored "-Wcovered-switch-default"
+#pragma GCC diagnostic ignored "-Wdeclaration-after-statement"
+#pragma GCC diagnostic ignored "-Wzero-as-null-pointer-constant"
+#ifndef __APPLE__
+#pragma GCC diagnostic ignored "-Wunsafe-buffer-usage"
```

Comment:
This is triggering a warning and needs a version check...

Playing around with https://godbolt.org/z/rxxovj9nd, it seems it was added in clang 16

Reply:
I added this, hopefully it's ok...

```
#ifndef __APPLE__
#if defined(__clang__) && __clang_major__ >= 16
#pragma GCC diagnostic ignored "-Wunsafe-buffer-usage"
#endif
#endif
```

Reply:
No wait that's not right! Should be `#if !defined(__clang__) || __clang_major__ >= 16`...